Refactor docsystem internals.

MichaelHatherly opened the pull request:
**Refactor docsystem internals**

The first commit consists of the following changes:

Structure of a module's `#meta` dict now has the following layout in all cases (with the exception of keywords):
- each module's `#meta` `ObjectIdDict` has keys of type `Binding` and values of type `MultiDoc`.
- each `MultiDoc` stores a collection of `DocStr` objects representing a single docstring each. These are ordered based on their order of definition rather than the current `type_morespecific` approach.
- `DocStr`s store the raw text of a docstring and lazily parse this to a `Markdown.MD` object when requested. By not parsing every docstring we also make some space savings in the `sys.*` files.

By keying `#meta` by `Binding` in every case the rest of the logic in `doc!` and `doc` becomes a lot more straightforward. `at-doc` expression building has also been simplified.

Additionally, the "summaries" displayed when no documentation can be found have been refactored and simplified.

**Rewrite genstdlib.jl**

The second commit takes advantage of the simplifications in the previous commit to reduce the complexity in this script. Outwardly the only difference should be the formatting of the summary that is displayed and speed (where timings drop from ~18 to ~9 seconds.)

MichaelHatherly committed:
Refactor docsystem internals

Structure of a module's `#meta` dict now has the following layout in
all cases (with the exception of keywords):

- each module's `#meta` `ObjectIdDict` has keys of type `Binding` and
  values of type `MultiDoc`.
- each `MultiDoc` stores a collection of `DocStr` objects representing a
  single docstring each. These are ordered based on their order of
  definition rather than the current `type_morespecific` approach.
- `DocStr`s store the raw text of a docstring and lazily parse this to a
  `Markdown.MD` object when requested. By not parsing every docstring we
  also make some space savings in the `sys.*` files.

By keying `#meta` by `Binding` in every case the rest of the logic in
`doc!` and `doc` becomes a lot more straightforward. `at-doc` expression
building has also been simplified.

Additionally, the "summaries" displayed when no documentation can be
found have been refactored and simplified.

MichaelHatherly committed:
Rewrite genstdlib.jl

This takes advantage of the simplifications in the previous commit to reduce
the complexity in this script. Outwardly the only difference should be
the formatting of the summary that is displayed and speed (where timings drop
from ~18 to ~9 seconds.)

tkelman commented on base/docs/Docs.jl:
```diff
@@ -244,165 +160,185 @@ is stored as `Union{Tuple{T}, Tuple{T, Any}}`.
+        println(io, "---\n")
+        println(io, readstring(readme))
+    else
+        println(io, "Additionally, no `README.md` found for module `", m, "`.\n")
```
what is this in addition to?

MichaelHatherly replied:
It was in addition to the default no docs message that appears, but it's not really necessary so I've remove the "Additionally".

tkelman commented on base/docs/Docs.jl:
```diff
@@ -430,40 +366,62 @@ nameof(other, ismacro)        = other
-    end
-end
+isfield(x) = isexpr(x, :.) &&
+  (isa(x.args[1], Symbol) || isfield(x.args[1])) &&
```
indentation

tkelman commented on base/docs/bootstrap.jl:
```diff
@@ -51,12 +51,15 @@ DocBootstrap
 """
 function loaddocs()
-    node = docs
+    node, tmp = docs, []
```
better variable name than `tmp`?

tkelman commented on doc/genstdlib.jl:
```diff
@@ -1,337 +1,198 @@
+    cd(root) do
+        state = State()
+        println("\n# GENSTDLIB\n")
+        # Find all available docstrings within `Base` and it's submodules.
```
its

tkelman commented:
Aside from little nitpicks, this does look like a really nice simplification. The fewer things we have to implicitly import into base to make the docsystem work, the better, and :100: for lazy parsing of the docstrings.

MichaelHatherly commented:
Thanks for reviewing @tkelman, I've pushed fixes for those comments.

MichaelHatherly committed:
Add metadata to markdown returned by `at-doc`.

This commit saves some of the values computed by `doc` and
`summarise` into the generated `MD` object that is returned.
These valus are the `Binding` and signature searched for, as
well as the vector of `DocStr` objects that match the `Binding`
and signature.

MichaelHatherly commented:
Found and fixed an issue in `parsedoc` where the generated markdown did not include the same metadata as it used to, namely a `:path` and `:module`.

Also added a small third commit that returns some extra metadata from `@doc` that will be useful when generating external docs. I'm happy to make a separate PR later for this if that would be preferred.

IainNZ commented on base/docs/Docs.jl:
```diff
@@ -244,165 +165,195 @@ is stored as `Union{Tuple{T}, Tuple{T, Any}}`.
+# =================
 
-# Modules
+function summarise(binding::Binding, sig)
```
Not sure the Americans will appreciate this name ;-)

MichaelHatherly replied:
Yeah, probably not :( Will fix.

MichaelHatherly replied:
Spellings changed in latest push.

tkelman commented:
CI failures were #15294, restarted
```
From worker 4:	     * numbers              Error During Test
	From worker 4:	  Test threw an exception of type UndefVarError
	From worker 4:	  Expression: round(T,true // false) === convert(T,Inf)
	From worker 4:	  UndefVarError: no_op_err not defined
	From worker 4:	   in gcd(::Bool, ::Bool) at ./intfuncs.jl:11
	From worker 4:	   in Rational{Bool}(::Bool, ::Bool) at ./rational.jl:9
	From worker 4:	   [inlined code] from /tmp/julia/share/julia/test/numbers.jl:2849
	From worker 4:	   in anonymous at ./no file:4294967295
	From worker 4:	   [inlined code] from ./essentials.jl:79
	From worker 4:	   in include_string(::UTF8String, ::ASCIIString) at ./loading.jl:371
	From worker 4:	   in include_from_node1(::ASCIIString) at ./loading.jl:420
	From worker 4:	   [inlined code] from ./util.jl:179
	From worker 4:	   in runtests(::ASCIIString) at /tmp/julia/share/julia/test/testdefs.jl:7
	From worker 4:	   in (::Base.Serializer.__deserialized_types__.##9137)(::ASCIIString) at /tmp/julia/share/julia/test/runtests.jl:36
	From worker 4:	   in run_work_thunk(::Base.##254#256{Base.CallMsg{:call_fetch}}, ::Bool) at ./multi.jl:714
	From worker 4:	   [inlined code] from ./multi.jl:1010
	From worker 4:	   in (::Base.##253#255{Base.CallMsg{:call_fetch},TCPSocket})() at ./task.jl:59
```

MichaelHatherly commented:
Barring further comments, I'll merge this later on today.
